### PR TITLE
feat: callback support #805

### LINKED ISSUE #805: Callback support for finetune recipe
**Is your feature request related to a problem? Please describe.**

We are integrating Customizer to use Automodel now for finetuning and need to support metrics support.

Previously, when we used NeMo for training, this was straightforward because NeMo is built on PyTorch Lightning, which has native callback support. We simply added a NeMoCustomizerCallback to report training progress to our API.

But with Automodel, my understanding is that it doesn't use PyTorch Lightning, so I can't just hook our callback. The simplest approach I can find is to subclass `[TrainFinetuneRecipeForNextTokenPrediction](https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/recipes/llm/train_ft.py#L853)` to override setup(), log_train_metrics(), log_val_metrics() methods to call our callback, but it doesn't seem to be the perfect solution.

**Describe the solution you'd like**

Could you add a callback mechanism similar to PyTorch Lightning's callbacks? Ideally with hooks for:

- on_train_start (after setup)
- on_train_batch_end (after each optimizer step)
- on_validation_end (after validation)
- on_save_checkpoint (when checkpoint is saved)
- on_exception (on training failure)

This would help us maintain cleaner integration with Customizer 

**Describe alternatives you've considered**
 
subclass `[TrainFinetuneRecipeForNextTokenPrediction](https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/recipes/llm/train_ft.py#L853)` to override setup(), log_train_metrics(), log_val_metrics() methods to call our callback

**Additional context**

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
Implements PyTorch Lightning-style callback system for Automodel recipes to enable custom integrations (e.g., Customizer metrics reporting, custom monitoring).

### Key Features
- **6 lifecycle hooks**: `on_train_start`, `on_train_batch_end`, `on_validation_end`, `on_save_checkpoint`, `on_exception`, `on_train_end`
- **`@rank_zero_only` decorator** for distributed training convenience
- **Full training context** passed to callbacks: recipe object, `MetricsSample` objects, checkpoint info, etc.
- **Programmatic API**: Callbacks passed directly to recipe constructors

### Changes
- Added `Callback` base class and `CallbackRunner` in `nemo_automodel/components/callbacks/`
- Integrated callbacks into all recipe classes (LLM finetune, VLM finetune, sequence classification, knowledge distillation)
- Comprehensive example: `examples/llm_finetune/finetune_with_callback.py`
- 9 unit tests covering all hooks and distributed behavior
- Documentation: `docs/guides/callbacks.md`

### Usage
```python
from nemo_automodel.components.callbacks import Callback

class MetricsReporter(Callback):
    def on_train_batch_end(self, recipe, **kwargs):
        log_data = kwargs['train_log_data']
        # Send metrics to external system
        
recipe = TrainFinetuneRecipeForNextTokenPrediction(
    cfg, 
    callbacks=[MetricsReporter()]
)
```

### Validation
**Unit tests**: All 9 tests pass ✅
```bash
python -m unittest tests.unit_tests.recipes.test_callbacks -v
Ran 9 tests in 0.006s
OK
```

**Example run**: 
```
python examples/llm_finetune/finetune_with_callback.py
```

Verified callbacks execute at correct lifecycle points with custom log prefixes:
```
2025-11-22 22:36:19 | INFO | __main__ | [SimpleLoggingCallback] 🔥 Training is starting!
2025-11-22 22:36:19 | INFO | __main__ | [SimpleLoggingCallback]    World size: 1 GPUs
2025-11-22 22:36:19 | INFO | __main__ | [SimpleLoggingCallback]    Total steps: 100
2025-11-22 22:37:33 | INFO | __main__ | [SimpleLoggingCallback] 🚀 Step 0/100: Loss = 0.9377, LR = 1.00e-05
...
2025-11-22 22:52:03 | INFO | __main__ | [SimpleLoggingCallback] ✅ Validation 'default': Loss = 0.2208
2025-11-22 22:52:03 | INFO | __main__ | [MetricsCollectorCallback] 📊 Collected 2 validation checkpoints
2025-11-22 22:52:03 | INFO | __main__ | [SimpleLoggingCallback] 💾 Checkpoint saved at step 99, epoch 0, path: checkpoints/epoch_0_step_99
2025-11-22 22:52:03 | INFO | __main__ | [MetricsCollectorCallback] 💾 Tracked checkpoint 2: step=99, train_loss=0.1678
2025-11-22 22:52:03 | INFO | __main__ | [SimpleLoggingCallback] 🎉 Training completed successfully! Final step: 100
2025-11-22 22:52:03 | INFO | __main__ | [MetricsCollectorCallback] 🎉 Training complete! Collected 100 training steps
```

Closes #805

---

Note: Happy to implement this callback feature! I designed it to be familiar (PyTorch Lightning-style) and practical for real integrations. Free free to provide feedback on the design/API if you'd like any adjustments. (AI assisted with documentation generation, but I personally reviewed and refined everything to ensure quality and accuracy.)

### CODE REVIEW

File: nemo_automodel/_cli/app.py [hunk 1]
```diff
@@ -232,6 +232,10 @@ def run_interactive(args):
     COMMAND_ALIASES = {"finetune": "train_ft", "pretrain": "train_ft"}
     # remap commands: finetune -> train_ft
     command = COMMAND_ALIASES.get(args.command, args.command)
+
+    # Special case for VLM finetune which uses finetune.py instead of train_ft.py
+    if args.domain == "vlm" and args.command == "finetune":
+        command = "finetune"
```
Comment (yuhezhang-ai) [inline nemo_automodel/_cli/app.py]: btw, I also fix a bug here; previusly the aliases would make the vlm finetune map to non-existing file

File: docs/guides/callbacks.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+recipe.setup()
+recipe.run_train_validation_loop()
+```
+
+## Distributed Training
+
+In multi-GPU training, callbacks run on **all ranks**. Python `logging` (e.g., `logger.info()`) is automatically filtered to rank 0, but other operations need explicit handling.
+
+### Using `@rank_zero_only` Decorator
```
Comment (jgerh) [inline docs/guides/callbacks.md]: We use imperative verbs instead of gerunds in headings

```suggestion
### Use `@rank_zero_only` Decorator
```

File: docs/guides/callbacks.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    @rank_zero_only
+    def on_validation_end(self, recipe, **kwargs):
+        # This only runs on rank 0
+        val_results = kwargs['val_results']
+        
+        # Safe to do file I/O, API calls, etc.
+        requests.post('https://api.example.com/metrics', json=val_results)
+```
+
+### Or Using Manual Rank Checking
```
Comment (jgerh) [inline docs/guides/callbacks.md]:
```suggestion
### Use Manual Rank Checking
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of .md file. Only two minor edits for our styles.

File: docs/guides/callbacks.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
+# Custom Callbacks
+
+## Introduction
+
+Callbacks provide a flexible way to inject custom logic into the training loop without modifying recipe code. They enable integration with external systems, custom logging, metrics collection, and monitoring.
```
Comment (jgerh) [inline docs/guides/callbacks.md:5]:
```suggestion
Callbacks provide a flexible way to inject custom logic into the training loop without modifying the recipe code. They enable integration with external systems, custom logging, metrics collection, and monitoring.
```

File: docs/guides/callbacks.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+## Complete Example
+
+See [`examples/llm_finetune/finetune_with_callback.py`](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/finetune_with_callback.py) for a full working example demonstrating:
+- Multiple callbacks
+- Distributed training with `@rank_zero_only`
+- Metrics collection for external reporting
+- Custom logging with prefixes
+
+### Running the Example
```
Comment (jgerh) [inline docs/guides/callbacks.md:142]:
```suggestion
### Run the Example
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review. Only 2 minor copyedits.
